recycle-bin.c

```diff
@@ -53,3 +49,69 @@ int wmain(int argc, wchar_t **argv) {
 
 	return ret;
 }
+
+int delete_files_new(int count, wchar_t **files) {
+	CHECK(CoInitializeEx(NULL, COINIT_MULTITHREADED));
+
+	IFileOperation *op;
+
+	CHECK(CoCreateInstance(
+		&CLSID_FileOperation,
+		NULL,
+		CLSCTX_ALL,
+		&IID_IFileOperation,
+		(void**)&op
+	));
+
+	CHECK_OBJ(op, op->lpVtbl->SetOperationFlags(
+		op,
+		FOFX_ADDUNDORECORD |
+		FOFX_RECYCLEONDELETE |
+		FOF_NOERRORUI |
+		FOF_SILENT |
+		FOFX_EARLYFAILURE
+	));
+
+	PCIDLIST_ABSOLUTE list[count - 1];
+
+	for (int i = 1; i < count; i++) {
```

Comment:
I found out that `IFileOperation` refuses to work with paths that are separated by forward slashes. This loop converts all paths from forward slashes to backward slashes.

Reply:
Is there really no system API to do this?

https://docs.microsoft.com/en-us/windows/desktop/FileIO/naming-a-file

Couldn't we make the paths "long paths" (`\\?\`-prefixed)? Those support forward slashes and they also don't have length restrictions.

Reply:
Also note:

> Starting in Windows 10, version 1607, MAX_PATH limitations have been removed from common Win32 file and directory functions. However, you must opt-in to the new behavior.
> You can also enable the new long path behavior per app via the manifest:

We should probably do that.

Reply:
I tried to prefix the paths with `\\?\` before passing them to `IFileOperation`, but apparently that doesn't work at all (not even for short paths). It seems like the Windows Shell API doesn't support long paths at all ([MSDN](https://social.msdn.microsoft.com/Forums/vstudio/en-US/42353ae3-1db9-45a2-a30f-99c184355b71/do-shell-objects-accept-long-path-names#6e53c00c-2d30-4692-a28e-253bbd6b04dd), [Stack Overflow](https://stackoverflow.com/questions/9980943/bypassing-max-path-limitation-for-itemidlist#comment12757715_9981012)). Although in theory it shouldn't accept paths longer than `MAX_PATH`, it still managed to delete a path with a length of over 400 characters in my Windows 7 VM 😄.

Reply:
> Starting in Windows 10, version 1607, MAX_PATH limitations have been removed from common Win32 file and directory functions. However, you must opt-in to the new behavior.
> You can also enable the new long path behavior per app via the manifest:

> These are the directory management functions that no longer have MAX_PATH restrictions if you opt-in to long path behavior: CreateDirectoryW, CreateDirectoryExW, ...

>These are the file management functions that no longer have MAX_PATH restrictions if you opt-in to long path behavior: CopyFileW, CopyFile2, ...

I think this only applies to the lower `Win32 File System` functions. The `Windows Shell API` is still [restricted](https://stackoverflow.com/questions/23976112/do-windows-shell-apis-work-with-long-unicode-paths).